dirty_cat/test/test_reproducibility.py

```diff
@@ -0,0 +1,23 @@
+import numpy as np
+
+from dirty_cat.similarity_encoder import SimilarityEncoder
+
+
+def get_result(random=None):
+    X = np.array(['aac', 'aaa', 'aaab', 'aaa', 'aaab', 'aaa', 'aaab', 'aaa']).reshape(-1, 1)
+    X_test = np.array([['Aa', 'aAa', 'aaa', 'aaab', ' aaa  c']]).reshape(-1, 1)
+    sim_enc = SimilarityEncoder(similarity='ngram', categories='auto', random_state=random)
+    encoder = sim_enc.fit(X)
+    transformed = encoder.transform(X_test)
+    return transformed
+
+
+def test_reproducibility():
+    res1 = get_result(2454)
+    res2 = get_result(2454)
+    assert (np.array_equal(res1, res2))
+
+    random_state = np.random.RandomState(32456)
+    res1 = get_result(random_state)
+    res2 = get_result(random_state)
+    assert (np.array_equal(res1, res2))
```

Comment:
Did these tests fail before? I am asking because the second part does not seem to test what it should: given that a stream is passed, different random streams are drawn in each execution. Hence, if the test passes, it's probably because it is not very stringent.

Reply:
Hint: to get a failure without seeding, you probably need a clustering problem that is not well posed, ie with string that all have the same distances in n-gram space. Try something like X = np.array([chr(i) for i in range(32, 127)]).reshape((-1, 1)). And I wouldn't bother separating X and X_test, I would just try sim_enc.fit_transform(X).

Reply:
Actually, I just tried that example, and it seems that it creates a very bad behavior in sim_enc, so we definitely need to test for it, and figure out what is wrong:
<pre>
X = np.array([chr(i) for i in range(32, 127)]).reshape((-1, 1))
sim_enc = SimilarityEncoder(categories='k-means', n_prototypes=10)
</pre>

Reply:
OK, I understood the problem with the above: the strings are so short that they all have an n-gram similarity of zero. Hence nothing woks.

Here is a suggested test, that fails without your fix:
<pre>
X = np.array([' %s ' % chr(i) for i in range(32, 127)]).reshape((-1, 1))
prototypes = sim_enc.fit(X).categories_
for i in range(10):
   assert (prototypes == sim_enc.fit(X).categories_[0]).all()
</pre>

Reply:
@GaelVaroquaux I pushed a fix. How ever I'm not really sure of what I did.
The only way I was able to get a positiv result on the case where we pass a RandomState is by doing a deep copy of it at every fit call, otherwise calling check_random_state when the type of random_state is int or None.

Reply:
Indeed, passing a random_state does not guarantee reproducibility, hence it should probably be removed from the test.